src/dbsync/src/sqlite_wrapper.cpp

```diff
@@ -13,147 +13,193 @@
+        };
+    }
+    return pDb;
 }
 
-bool Connection::Close() {
-  auto ret_val { false };
-  
-  if (nullptr != m_db_instance) {
-    ret_val = SQLITE_OK == sqlite3_close_v2(m_db_instance);
-    m_db_instance = nullptr;
-  }
+Connection::Connection(const std::string& path)
+: m_db{ openSQLiteDb(path), [](sqlite3* p){ sqlite3_close_v2(p); } }
+{}
 
-  return ret_val;
-}
-sqlite3* Connection::GetDBInstance() {
-  return m_db_instance;
+bool Connection::close()
+{
+    m_db.reset();
+    return !!m_db;
 }
 
-Connection::Connection() : Connection(kDefaultPath){
+const std::shared_ptr<sqlite3>& Connection::db() const
+{
+    return m_db;
 }
 
-bool Connection::Execute(const std::string& query) {
-  auto ret_val{ false };
-  if (nullptr != m_db_instance) {
-    ret_val = SQLITE_OK == sqlite3_exec(m_db_instance, query.c_str(), 0,0, nullptr);
-  }
-  return ret_val;
-}
+Connection::Connection()
+: Connection(DB_DEFAULT_PATH)
+{}
 
-Transaction::~Transaction() {
-  if (!m_rollbacked && !m_commited) {
-    m_connection->Execute("ROLLBACK TRANSACTION");
-  }
+bool Connection::execute(const std::string& query)
+{
+    return m_db &&
+           SQLITE_OK == sqlite3_exec(m_db.get(), query.c_str(), 0,0, nullptr);
 }
 
-Transaction::Transaction(std::shared_ptr<IConnection>& connection) : m_connection(connection){
-  m_started = false;
-  m_commited = false;
-  m_rollbacked = false;
+Transaction::~Transaction()
+{
+    if (!m_rolledBack && !m_commited)
+    {
+        m_connection->execute("ROLLBACK TRANSACTION");
+    }
+}
 
-  if (m_connection->Execute("BEGIN TRANSACTION")) {
-    m_started = true;
-  }
+Transaction::Transaction(std::shared_ptr<IConnection>& connection)
+: m_connection{ connection }
+, m_rolledBack{ false }
+, m_commited{ false }
+{
+    if (!m_connection->execute("BEGIN TRANSACTION"))
+    {
+        throw SQLite::exception
+        {
+            601,
+            "cannot begin SQLite Transaction."
+        };
+    }
 }
     
-bool Transaction::Commit() {
-  if (m_started && 
-  !m_rollbacked && 
-  !m_commited) 
-  {
-    if (m_connection->Execute("COMMIT TRANSACTION")) {
-      m_commited = true;
+bool Transaction::commit()
+{
+    if (!m_rolledBack && !m_commited)
+    {
+        m_commited = m_connection->execute("COMMIT TRANSACTION");
     }
-  }
-  return m_commited;
-}
-
-bool Transaction::Rollback() {
-  if (m_started && 
-  !m_rollbacked && 
-  !m_commited) 
-  {
-    m_connection->Execute("ROLLBACK TRANSACTION");
-    m_rollbacked = true;
-  }
-  return m_rollbacked;
+    return m_commited;
 }
 
-
-Statement::Statement(std::shared_ptr<IConnection>& connection, const std::string& query) : m_connection(connection){
-  if(SQLITE_OK != sqlite3_prepare_v2(connection->GetDBInstance(), query.c_str(), -1, &m_stmt, nullptr)) {
-    throw SQLite::exception(601, "cannot instance SQLite stmt.");
-  }
+bool Transaction::rollback()
+{
+    if (!m_rolledBack && !m_commited)
+    {
+        m_connection->execute("ROLLBACK TRANSACTION");
```

Comment:
shouldn't the "execute" return value needed to be checked before going on?

Reply:
if rollback fails there not much left to do and destructors must not throw exceptions.